duck-tape-android/src/main/java/leakcanary/DuckTapeAndroid.kt

```diff
@@ -0,0 +1,240 @@
+   * an OnClickListener on it and then never recycles it, expecting it to be garbage collected
+   * but it ends up being held by the GoogleApiHandler thread.
+   */
+  GOOGLE_API_HANDLER_THREAD {
+    override fun apply(application: Application) {
+      // 30 seconds is an arbitrary delay to give time to Play Services to start the thread.
+      backgroundExecutor.schedule({
+        val thread = findGoogleApiHandlerThread()
+        if (thread != null) {
+          // Unfortunately Looper.getQueue() is API 23. Looper.myQueue() is API 1.
+          // So we have to post to the handler thread to be able to obtain the queue for that
+          // thread from within that Thread.
+          // When the Google API thread becomes idle, we post a message to force it to move.
+          // Source: https://developer.squareup.com/blog/a-small-leak-will-sink-a-great-ship/
+          val flushHandler = Handler(thread.looper)
+          flushHandler.post {
+            Looper.myQueue()
+                .addIdleHandler {
+                  flushHandler.sendMessageDelayed(flushHandler.obtainMessage(), 1000)
+                  true
+                }
+          }
+        }
+      }, 30, TimeUnit.SECONDS)
+    }
+
+    private fun findGoogleApiHandlerThread(): HandlerThread? {
+      // https://stackoverflow.com/a/1323480
+      var rootGroup = Thread.currentThread()
+          .threadGroup
+      var lookForRoot = true
+      while (lookForRoot) {
+        val parentGroup = rootGroup.parent
+        if (parentGroup != null) {
+          rootGroup = parentGroup
+        } else {
+          lookForRoot = false
+        }
+      }
+      var threads = arrayOfNulls<Thread>(rootGroup.activeCount())
+      while (rootGroup.enumerate(threads, true) == threads.size) {
+        threads = arrayOfNulls(threads.size * 2)
+      }
+      for (thread in threads) {
+        if (thread is HandlerThread && thread.getName() == "GoogleApiHandler") {
+          return thread
+        }
+      }
+      return null
+    }
+  },
+
+  /**
+   * Until API 28, AccessibilityNodeInfo has a mOriginalText field that was not properly cleared
+   * when instance were put back in the pool.
+   * Leak introduced here: https://android.googlesource.com/platform/frameworks/base/+
+   * /193520e3dff5248ddcf8435203bf99d2ba667219%5E%21/core/java/android/view/accessibility
+   * /AccessibilityNodeInfo.java
+   *
+   * Fixed here: https://android.googlesource.com/platform/frameworks/base/+
+   * /6f8ec1fd8c159b09d617ed6d9132658051443c0c
+   */
+  ACCESSIBILITY_NODE_INFO {
+    override fun apply(application: Application) {
+      if (SDK_INT >= 28) {
+        return
+      }
+      // Starve the pool on activity destroy
+      application.onActivityDestroyed {
+        val maxPoolSize = 50
+        for (i in 0 until maxPoolSize) {
+          AccessibilityNodeInfo.obtain()
+        }
+      }
+    }
+  }
+
+  ;
```

Comment:
What is this `;` ?

Reply:
Required `;` after the last element of an enum.